# **Studi Kasus Kredit Score (Tugas 7)**



In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder
from sklearn.compose import make_column_transformer

### Insert Dataset

In [2]:
  data = pd.read_csv("https://raw.githubusercontent.com/masaul/data-csv/main/credit_score.csv")
  data

Unnamed: 0 kode_kontrak  pendapatan_setahun_juta kpr_aktif  \
0             1   AGR-000001                      295        YA   
1             2   AGR-000011                      271        YA   
2             3   AGR-000030                      159     TIDAK   
3             4   AGR-000043                      210        YA   
4             5   AGR-000049                      165     TIDAK   
..          ...          ...                      ...       ...   
895         896   AGR-010739                      112        YA   
896         897   AGR-010744                      120        YA   
897         898   AGR-010758                      166     TIDAK   
898         899   AGR-010775                      196     TIDAK   
899         900   AGR-010790                      204     TIDAK   

     durasi_pinjaman_bulan  jumlah_tanggungan rata_rata_overdue  risk_rating  
0                       48                  5      61 - 90 days            4  
1                       36                  5      61 - 90 days            4  
2                       12                  0       0 - 30 days            1  
3                       12                  3      46 - 60 days            3  
4                       36                  0      31 - 45 days            2  
..                     ...                ...               ...          ...  
895                     48                  5         > 90 days            5  
896                     48                  2      46 - 60 days            3  
897                     24                  2       0 - 30 days            1  
898                     48                  0      31 - 45 days            2  
899                     24                  2       0 - 30 days            1  

[900 rows x 8 columns]

### Convert ke dataframe

In [3]:
df_without_column_for_convert = pd.DataFrame(data, columns = ['kode_kontrak','pendapatan_setahun_juta','durasi_pinjaman_bulan','jumlah_tanggungan','risk_rating'])
df_without_column_for_convert.head()

kode_kontrak  pendapatan_setahun_juta  durasi_pinjaman_bulan  \
0   AGR-000001                      295                     48   
1   AGR-000011                      271                     36   
2   AGR-000030                      159                     12   
3   AGR-000043                      210                     12   
4   AGR-000049                      165                     36   

   jumlah_tanggungan  risk_rating  
0                  5            4  
1                  5            4  
2                  0            1  
3                  3            3  
4                  0            2

### Preprocessing
Mengubah data categorial ke numerik

In [4]:
df_ratarata=pd.get_dummies(data['rata_rata_overdue'])
df_ratarata.head()

0 - 30 days  31 - 45 days  46 - 60 days  61 - 90 days  > 90 days
0            0             0             0             1          0
1            0             0             0             1          0
2            1             0             0             0          0
3            0             0             1             0          0
4            0             1             0             0          0

In [5]:
df_kpr_aktif=pd.get_dummies(data['kpr_aktif'])
df_kpr_aktif

TIDAK  YA
0        0   1
1        0   1
2        1   0
3        0   1
4        1   0
..     ...  ..
895      0   1
896      0   1
897      1   0
898      1   0
899      1   0

[900 rows x 2 columns]

In [6]:
df_new = pd.concat([df_without_column_for_convert, df_kpr_aktif,df_ratarata], axis=1)
df_new.head()

kode_kontrak  pendapatan_setahun_juta  durasi_pinjaman_bulan  \
0   AGR-000001                      295                     48   
1   AGR-000011                      271                     36   
2   AGR-000030                      159                     12   
3   AGR-000043                      210                     12   
4   AGR-000049                      165                     36   

   jumlah_tanggungan  risk_rating  TIDAK  YA  0 - 30 days  31 - 45 days  \
0                  5            4      0   1            0             0   
1                  5            4      0   1            0             0   
2                  0            1      1   0            1             0   
3                  3            3      0   1            0             0   
4                  0            2      1   0            0             1   

   46 - 60 days  61 - 90 days  > 90 days  
0             0             1          0  
1             0             1          0  
2             0             0          0  
3             1             0          0  
4             0             0          0

In [7]:
df_risk_rating = pd.DataFrame(data, columns = ['risk_rating'])
df_risk_rating.head()

risk_rating
0            4
1            4
2            1
3            3
4            2

In [8]:
df_without_risk_rating=df_new.drop(['risk_rating'], axis=1)

In [9]:
df_new2 = pd.concat([df_without_risk_rating,df_risk_rating], axis=1)
df_new2

kode_kontrak  pendapatan_setahun_juta  durasi_pinjaman_bulan  \
0     AGR-000001                      295                     48   
1     AGR-000011                      271                     36   
2     AGR-000030                      159                     12   
3     AGR-000043                      210                     12   
4     AGR-000049                      165                     36   
..           ...                      ...                    ...   
895   AGR-010739                      112                     48   
896   AGR-010744                      120                     48   
897   AGR-010758                      166                     24   
898   AGR-010775                      196                     48   
899   AGR-010790                      204                     24   

     jumlah_tanggungan  TIDAK  YA  0 - 30 days  31 - 45 days  46 - 60 days  \
0                    5      0   1            0             0             0   
1                    5      0   1            0             0             0   
2                    0      1   0            1             0             0   
3                    3      0   1            0             0             1   
4                    0      1   0            0             1             0   
..                 ...    ...  ..          ...           ...           ...   
895                  5      0   1            0             0             0   
896                  2      0   1            0             0             1   
897                  2      1   0            1             0             0   
898                  0      1   0            0             1             0   
899                  2      1   0            1             0             0   

     61 - 90 days  > 90 days  risk_rating  
0               1          0            4  
1               1          0            4  
2               0          0            1  
3               0          0            3  
4               0          0            2  
..            ...        ...          ...  
895             0          1            5  
896             0          0            3  
897             0          0            1  
898             0          0            2  
899             0          0            1  

[900 rows x 12 columns]

### Normalisasi

In [10]:
from sklearn.preprocessing import MinMaxScaler
df_for_minmax_scaler=pd.DataFrame(data, columns = ['pendapatan_setahun_juta','durasi_pinjaman_bulan','jumlah_tanggungan'])
df_for_minmax_scaler.to_numpy()
scaler = MinMaxScaler()
df_hasil_minmax_scaler=scaler.fit_transform(df_for_minmax_scaler)

In [11]:
df_hasil_minmax_scaler = pd.DataFrame(df_hasil_minmax_scaler,columns = ['pendapatan_setahun_juta','durasi_pinjaman_bulan','jumlah_tanggungan'])
df_hasil_minmax_scaler

pendapatan_setahun_juta  durasi_pinjaman_bulan  jumlah_tanggungan
0                   0.978261               1.000000           0.833333
1                   0.873913               0.666667           0.833333
2                   0.386957               0.000000           0.000000
3                   0.608696               0.000000           0.500000
4                   0.413043               0.666667           0.000000
..                       ...                    ...                ...
895                 0.182609               1.000000           0.833333
896                 0.217391               1.000000           0.333333
897                 0.417391               0.333333           0.333333
898                 0.547826               1.000000           0.000000
899                 0.582609               0.333333           0.333333

[900 rows x 3 columns]

In [12]:
df_drop_column_for_minmaxscaler=df_new2.drop(['pendapatan_setahun_juta','durasi_pinjaman_bulan','jumlah_tanggungan'], axis=1)
df_drop_column_for_minmaxscaler

kode_kontrak  TIDAK  YA  0 - 30 days  31 - 45 days  46 - 60 days  \
0     AGR-000001      0   1            0             0             0   
1     AGR-000011      0   1            0             0             0   
2     AGR-000030      1   0            1             0             0   
3     AGR-000043      0   1            0             0             1   
4     AGR-000049      1   0            0             1             0   
..           ...    ...  ..          ...           ...           ...   
895   AGR-010739      0   1            0             0             0   
896   AGR-010744      0   1            0             0             1   
897   AGR-010758      1   0            1             0             0   
898   AGR-010775      1   0            0             1             0   
899   AGR-010790      1   0            1             0             0   

     61 - 90 days  > 90 days  risk_rating  
0               1          0            4  
1               1          0            4  
2               0          0            1  
3               0          0            3  
4               0          0            2  
..            ...        ...          ...  
895             0          1            5  
896             0          0            3  
897             0          0            1  
898             0          0            2  
899             0          0            1  

[900 rows x 9 columns]

In [13]:
df_new3 = pd.concat([df_drop_column_for_minmaxscaler,df_hasil_minmax_scaler], axis=1)
df_new3

kode_kontrak  TIDAK  YA  0 - 30 days  31 - 45 days  46 - 60 days  \
0     AGR-000001      0   1            0             0             0   
1     AGR-000011      0   1            0             0             0   
2     AGR-000030      1   0            1             0             0   
3     AGR-000043      0   1            0             0             1   
4     AGR-000049      1   0            0             1             0   
..           ...    ...  ..          ...           ...           ...   
895   AGR-010739      0   1            0             0             0   
896   AGR-010744      0   1            0             0             1   
897   AGR-010758      1   0            1             0             0   
898   AGR-010775      1   0            0             1             0   
899   AGR-010790      1   0            1             0             0   

     61 - 90 days  > 90 days  risk_rating  pendapatan_setahun_juta  \
0               1          0            4                 0.978261   
1               1          0            4                 0.873913   
2               0          0            1                 0.386957   
3               0          0            3                 0.608696   
4               0          0            2                 0.413043   
..            ...        ...          ...                      ...   
895             0          1            5                 0.182609   
896             0          0            3                 0.217391   
897             0          0            1                 0.417391   
898             0          0            2                 0.547826   
899             0          0            1                 0.582609   

     durasi_pinjaman_bulan  jumlah_tanggungan  
0                 1.000000           0.833333  
1                 0.666667           0.833333  
2                 0.000000           0.000000  
3                 0.000000           0.500000  
4                 0.666667           0.000000  
..                     ...                ...  
895               1.000000           0.833333  
896               1.000000           0.333333  
897               0.333333           0.333333  
898               1.000000           0.000000  
899               0.333333           0.333333  

[900 rows x 12 columns]

In [14]:
df_new3_without_risk_rating=df_new3.drop(['risk_rating'], axis=1)
df_new3_without_risk_rating

kode_kontrak  TIDAK  YA  0 - 30 days  31 - 45 days  46 - 60 days  \
0     AGR-000001      0   1            0             0             0   
1     AGR-000011      0   1            0             0             0   
2     AGR-000030      1   0            1             0             0   
3     AGR-000043      0   1            0             0             1   
4     AGR-000049      1   0            0             1             0   
..           ...    ...  ..          ...           ...           ...   
895   AGR-010739      0   1            0             0             0   
896   AGR-010744      0   1            0             0             1   
897   AGR-010758      1   0            1             0             0   
898   AGR-010775      1   0            0             1             0   
899   AGR-010790      1   0            1             0             0   

     61 - 90 days  > 90 days  pendapatan_setahun_juta  durasi_pinjaman_bulan  \
0               1          0                 0.978261               1.000000   
1               1          0                 0.873913               0.666667   
2               0          0                 0.386957               0.000000   
3               0          0                 0.608696               0.000000   
4               0          0                 0.413043               0.666667   
..            ...        ...                      ...                    ...   
895             0          1                 0.182609               1.000000   
896             0          0                 0.217391               1.000000   
897             0          0                 0.417391               0.333333   
898             0          0                 0.547826               1.000000   
899             0          0                 0.582609               0.333333   

     jumlah_tanggungan  
0             0.833333  
1             0.833333  
2             0.000000  
3             0.500000  
4             0.000000  
..                 ...  
895           0.833333  
896           0.333333  
897           0.333333  
898           0.000000  
899           0.333333  

[900 rows x 11 columns]

### Data yang akan digunakan

In [15]:
df_new4 = pd.concat([df_new3_without_risk_rating,df_risk_rating], axis=1)
df_new4

kode_kontrak  TIDAK  YA  0 - 30 days  31 - 45 days  46 - 60 days  \
0     AGR-000001      0   1            0             0             0   
1     AGR-000011      0   1            0             0             0   
2     AGR-000030      1   0            1             0             0   
3     AGR-000043      0   1            0             0             1   
4     AGR-000049      1   0            0             1             0   
..           ...    ...  ..          ...           ...           ...   
895   AGR-010739      0   1            0             0             0   
896   AGR-010744      0   1            0             0             1   
897   AGR-010758      1   0            1             0             0   
898   AGR-010775      1   0            0             1             0   
899   AGR-010790      1   0            1             0             0   

     61 - 90 days  > 90 days  pendapatan_setahun_juta  durasi_pinjaman_bulan  \
0               1          0                 0.978261               1.000000   
1               1          0                 0.873913               0.666667   
2               0          0                 0.386957               0.000000   
3               0          0                 0.608696               0.000000   
4               0          0                 0.413043               0.666667   
..            ...        ...                      ...                    ...   
895             0          1                 0.182609               1.000000   
896             0          0                 0.217391               1.000000   
897             0          0                 0.417391               0.333333   
898             0          0                 0.547826               1.000000   
899             0          0                 0.582609               0.333333   

     jumlah_tanggungan  risk_rating  
0             0.833333            4  
1             0.833333            4  
2             0.000000            1  
3             0.500000            3  
4             0.000000            2  
..                 ...          ...  
895           0.833333            5  
896           0.333333            3  
897           0.333333            1  
898           0.000000            2  
899           0.333333            1  

[900 rows x 12 columns]

## Implementasi Metode Naive Bayes

In [16]:
from sklearn.metrics import make_scorer, accuracy_score,precision_score
from sklearn.metrics import accuracy_score ,precision_score,recall_score,f1_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import KFold,train_test_split,cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split

In [17]:
X=df_new4.iloc[:,1:10].values
y=df_new4.iloc[:,11].values

In [18]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

In [19]:
gaussian = GaussianNB()
gaussian.fit(X_train, y_train)
Y_pred = gaussian.predict(X_test) 
accuracy_nb=round(accuracy_score(y_test,Y_pred)* 100, 2)
acc_gaussian = round(gaussian.score(X_train, y_train) * 100, 2)

cm = confusion_matrix(y_test, Y_pred)
accuracy = accuracy_score(y_test,Y_pred)
precision =precision_score(y_test, Y_pred,average='micro')
recall =  recall_score(y_test, Y_pred,average='micro')
f1 = f1_score(y_test,Y_pred,average='micro')
print('Confusion matrix for Naive Bayes\n',cm)
print('accuracy_Naive Bayes: %.3f' %accuracy)
print('precision_Naive Bayes: %.3f' %precision)
print('recall_Naive Bayes: %.3f' %recall)
print('f1-score_Naive Bayes : %.3f' %f1)

Confusion matrix for Naive Bayes
 [[69  0  0  0  0]
 [ 0 49  0  0  0]
 [ 0  0 84  0  0]
 [ 0  0  0 36  0]
 [ 0  0  0  0 32]]
accuracy_Naive Bayes: 1.000
precision_Naive Bayes: 1.000
recall_Naive Bayes: 1.000
f1-score_Naive Bayes : 1.000


## Implementasi Metode K-NN

In [20]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 25)
knn.fit(X_train,y_train)
Y_pred = knn.predict(X_test)[0:]
accuracy_nb=round(accuracy_score(y_test,Y_pred)* 100, 2)
acc_neigh = round(knn.score(X_train, y_train) * 100, 2)

cm = confusion_matrix(y_test, Y_pred)
accuracy = accuracy_score(y_test,Y_pred)
precision =precision_score(y_test, Y_pred,average='micro')
recall =  recall_score(y_test, Y_pred,average='micro')
f1 = f1_score(y_test,Y_pred,average='micro')
print('Confusion matrix for K-NN\n',cm)
print('accuracy_K-NN: %.3f' %accuracy)
print('precision_K-NN: %.3f' %precision)
print('recall_K-NN: %.3f' %recall)
print('f1-score_K-NN : %.3f' %f1)

Confusion matrix for K-NN
 [[69  0  0  0  0]
 [ 0 49  0  0  0]
 [ 0  1 83  0  0]
 [ 0  0  0 36  0]
 [ 0  1  0  0 31]]
accuracy_K-NN: 0.993
precision_K-NN: 0.993
recall_K-NN: 0.993
f1-score_K-NN : 0.993


## Implementasi Metode Decision Tree

In [21]:
from sklearn import tree

iristree = tree.DecisionTreeClassifier(max_depth=3, criterion='gini', random_state=0)
iristree.fit(X_train, y_train)
Y_pred = iristree.predict(X_test) 
accuracy_nb=round(accuracy_score(y_test,Y_pred)* 100, 2)
acc_gaussian = round(gaussian.score(X_train, y_train) * 100, 2)

cm = confusion_matrix(y_test, Y_pred)
accuracy = accuracy_score(y_test,Y_pred)
precision =precision_score(y_test, Y_pred,average='micro')
recall =  recall_score(y_test, Y_pred,average='micro')
f1 = f1_score(y_test,Y_pred,average='micro')
print('Confusion matrix for Decision Tree\n',cm)
print('accuracy_Decision Tree: %.3f' %accuracy)
print('precision_Decision Tree: %.3f' %precision)
print('recall_Decision Tree: %.3f' %recall)
print('f1-score_Decision Tree : %.3f' %f1)


Confusion matrix for Decision Tree
 [[69  0  0  0  0]
 [ 0 49  0  0  0]
 [ 0  0 84  0  0]
 [ 0  0  0 36  0]
 [ 0  0  0 32  0]]
accuracy_Decision Tree: 0.881
precision_Decision Tree: 0.881
recall_Decision Tree: 0.881
f1-score_Decision Tree : 0.881


Hasil Akurasi yang didapatkan:


1.   Metode Naive Bayes hasilnya 100%
2.   Metode K-NN hasilnya 99%
3.   Metode Decision Tree hasilnya 88%

